plugins/derive/derive.ml

```diff
@@ -30,15 +29,17 @@ let start_deriving f suchthat name : Declare.Proof.t =
   let goals =
     let open Proofview in
     TCons ( env , sigma , f_type_type , (fun sigma f_type ->
+        let sigma = match typ with
+          | None -> sigma
+          | Some typ ->
+            let sigma, typ = Constrintern.interp_type_evars ~program_mode:false env sigma typ in
+            Evd.define (fst (EConstr.destEvar sigma f_type)) typ sigma in
```

Comment:
Frankly, this looks overly complicated  doesn't it? Wouldn't it be more straightforward to not create the first goal at all when `typ` is specified? Rather than creating a goal and immediately solve it?

Reply:
Hi @aspiwack, glad to hear from you!

I wanted to do minimal changes. In #19092, I actually renounce to pass the type of the witness to the telescope, and to rely on the ability to insert computation in `Proof.dependent_start` because I need to get back the signature of implicit arguments which would I did not see a way to get back if done in `Proof.dependent_start`.

Reply:
Note an issue of `Proof.start` and `Proof.dependent_start` about which I didn't figure what was the intent: if the original sigma passed to `Proof.start` has undefined evars, where are these evars supposed to live? They are not put in the shelf, so I concluded that we have to manually shelve them before calling `Proof.start` (see this [line](https://github.com/coq/coq/blob/9c84f76b50087300e831141ebd91bff0fd67fc1e/plugins/derive/derive.ml#L32) in the PR). Is that correct? Or should alternatively `Proof.start` do this shelving itself?

Reply:
As you can see in #19092, thanks to Emilio's work in 2020 on centralizing all declarations in `declare.ml`, the perspective of using the proof engine to support a general and unified system of declarations supporting arbitrarily mutual, dependent, and non-necessarily evar-free statements (as in `Derive`) gets closer.

Reply:
> Hi @aspiwack, glad to hear from you!

I'm always lurking :slightly_smiling_face: .

> I wanted to do minimal changes.

This is fair enough.

> Note an issue of `Proof.start` and `Proof.dependent_start` about which I didn't figure what was the intent: if the original sigma passed to `Proof.start` has undefined evars, where are these evars supposed to live?

I can't say that I remember. It's been a while. I'm guessing that I didn't document this very well, did I? I vaguely remember that there was some special treatment of the existential variables that weren't in the shelf. Maybe that's what `future_goals` field was about?

Reply:
Yes, future goals are collecting evars produced while refining. But the thing I'm asking about is about the proof initialization. It is not really a worry at the current time (as we can shelve ourselves before initializing). If ever needed, we'll take a decision. Thanks!